File: scripts/import-data.ts
import { PrismaClient } from '@prisma/client';
import { parse } from 'csv-parse';
import * as fs from 'fs';
import * as path from 'path';

const prisma = new PrismaClient();

interface QuestionData {
  type: string;
  options: string[];
  question: string;
}

async function parseCSV(filePath: string): Promise<any[]> {
  const records: any[] = [];
  const parser = fs
    .createReadStream(filePath)
    .pipe(parse({ columns: true, skip_empty_lines: true }));

  for await (const record of parser) {
    records.push(record);
  }
  return records;
}

async function importQuestionnaires() {
  const records = await parseCSV(path.join(__dirname, '../data/questionnaire_questionnaires.csv'));
  
  for (const record of records) {
    await prisma.questionnaire.create({
      data: {
        id: parseInt(record.id),
        name: record.name,
      },
    });
  }
  console.log('Questionnaires imported successfully');
}

async function importQuestions() {
  const records = await parseCSV(path.join(__dirname, '../data/questionnaire_questions.csv'));
  
  for (const record of records) {
    const questionData: QuestionData = JSON.parse(record.question);
    await prisma.question.create({
      data: {
        id: parseInt(record.id),
        type: questionData.type,
        options: JSON.stringify(questionData.options || []),
        question: questionData.question,
      },
    });
  }
  console.log('Questions imported successfully');
}

async function importJunctions() {
  const records = await parseCSV(path.join(__dirname, '../data/questionnaire_junction.csv'));
  
  for (const record of records) {
    await prisma.questionJunction.create({
      data: {
        id: parseInt(record.id),
        questionId: parseInt(record.question_id),
        questionnaireId: parseInt(record.questionnaire_id),
        priority: parseInt(record.priority),
      },
    });
  }
  console.log('Question junctions imported successfully');
}

async function createAdminUser() {
  const adminUsername = process.env.ADMIN_USERNAME || 'admin';
  const adminPassword = process.env.ADMIN_PASSWORD || 'admin123';

  await prisma.user.create({
    data: {
      username: adminUsername,
      password: adminPassword, // In a real app, this should be hashed
      isAdmin: true,
    },
  });
  console.log('Admin user created successfully');
}

async function main() {
  try {
    // Clear existing data
    await prisma.$transaction([
      prisma.answer.deleteMany(),
      prisma.response.deleteMany(),
      prisma.questionJunction.deleteMany(),
      prisma.question.deleteMany(),
      prisma.questionnaire.deleteMany(),
      prisma.user.deleteMany(),
    ]);
    console.log('Existing data cleared');

    // Import data in order
    await importQuestionnaires();
    await importQuestions();
    await importJunctions();
    await createAdminUser();

    console.log('Data import completed successfully');
  } catch (error) {
    console.error('Error importing data:', error);
    process.exit(1);
  } finally {
    await prisma.$disconnect();
  }
}

main();

File: backend/app/import_data.py
import pandas as pd
import json
import uuid
from sqlalchemy.orm import Session
from . import models
from .database import SessionLocal, engine

def import_data():
    # Create tables
    models.Base.metadata.create_all(bind=engine)
    
    db = SessionLocal()
    try:
        # Clear existing data
        db.query(models.Answer).delete()
        db.query(models.Response).delete()
        db.query(models.QuestionJunction).delete()
        db.query(models.Question).delete()
        db.query(models.Questionnaire).delete()
        db.query(models.User).delete()
        
        # Import questionnaires
        questionnaires_df = pd.read_csv('../data/questionnaire_questionnaires.csv')
        for _, row in questionnaires_df.iterrows():
            questionnaire = models.Questionnaire(
                id=row['id'],
                name=row['name']
            )
            db.add(questionnaire)
        
        # Import questions
        questions_df = pd.read_csv('../data/questionnaire_questions.csv')
        for _, row in questions_df.iterrows():
            question_data = json.loads(row['question'])
            question = models.Question(
                id=row['id'],
                type=question_data['type'],
                options=question_data.get('options', []),
                question=question_data['question']
            )
            db.add(question)
        
        # Import junctions
        junctions_df = pd.read_csv('../data/questionnaire_junction.csv')
        for _, row in junctions_df.iterrows():
            junction = models.QuestionJunction(
                id=row['id'],
                question_id=row['question_id'],
                questionnaire_id=row['questionnaire_id'],
                priority=row['priority']
            )
            db.add(junction)
        
        # Create admin user
        admin = models.User(
            id=str(uuid.uuid4()),
            username='admin',
            password='admin123',  # In production, this should be hashed
            is_admin=True
        )
        db.add(admin)
        
        db.commit()
        print("Data import completed successfully")
        
    except Exception as e:
        print(f"Error importing data: {e}")
        db.rollback()
        raise
    finally:
        db.close()

if __name__ == "__main__":
    import_data()
